src/components/smart/Table.vue

```diff
@@ -1,69 +1,321 @@
+<script lang="ts" setup>
+import { useVModel } from "@vueuse/core"
+import { isEqual } from "lodash-es"
+import { computed, ref, watch } from "vue"
+import IconLeft from "~icons/lucide/arrow-left"
+import IconRight from "~icons/lucide/arrow-right"
+
 export type CellHeading = {
   key: string
   label?: string
   preventClick?: boolean
 }
 
-defineProps<{
-  /** Whether to show the vertical border between columns */
-  showYBorder?: boolean
-  /**  The list of items to be displayed in the table */
-  list?: Item[]
-  /** The headings of the table */
-  headings?: CellHeading[]
-}>()
+export type Item = Record<string, unknown>
+
+const props = withDefaults(
+  defineProps<{
+    /** Whether to show the vertical border between columns */
+    showYBorder?: boolean
+    /**  The list of items to be displayed in the table */
+    list: Item[]
+    /** The headings of the table */
+    headings?: CellHeading[]
+    /** Whether to show the search bar */
+    searchBar?: {
+      /** Whether to debounce the search query event */
+      debounce?: number
+      placeholder?: string
+    }
+    /** Whether to show the checkbox column
+     * This will be overriden if custom implementation for body slot is provided
+     */
+    checkbox?: boolean
+
+    selectedRows?: Item[]
+    /** Whether to enable sorting */
+    sort?: {
+      /** The key to sort the list by */
+      key: string
+      direction: Direction
+    }
+
+    /** Whether to enable pagination */
+    pagination?: {
+      totalPages: number
+    }
+
+    /** Whether to show the spinner */
+    spinner?: {
+      enabled: boolean
+      duration?: number
+    }
+  }>(),
+  {
+    showYBorder: false,
+    search: undefined,
+    checkbox: false,
+    sort: undefined,
+    selectedRows: undefined,
+  },
+)
 
 const emit = defineEmits<{
   (event: "onRowClicked", item: Item): void
+  (event: "update:list", list: Item[]): void
+  (event: "search", query: string): void
+  (event: "update:selectedRows", selectedRows: Item[]): void
+  (event: "pageNumber", page: number): void
 }>()
 
+// Pagination functionality
+const page = ref(1)
+
+enum PageDirection {
+  Previous,
+  Next,
+}
+
+const changePage = (direction: PageDirection) => {
+  const isPrevious = direction === PageDirection.Previous
+  if (
+    (isPrevious && page.value > 1) ||
+    (!isPrevious && page.value < props.pagination!.totalPages)
+  ) {
+    page.value += isPrevious ? -1 : 1
+  }
+
+  emit("pageNumber", page.value)
+}
+
+// The working version of the list that is used to perform operations upon
+const workingList = useVModel(props, "list", emit)
+
+// Spinner functionality
+const isSpinnerEnabled = ref(false)
+const showSpinner = (duration: number = 500) => {
+  isSpinnerEnabled.value = true
+  setTimeout(() => {
+    isSpinnerEnabled.value = false
+  }, duration)
+}
+
+watch(
+  () => props.spinner,
+  () => {
+    if (props.spinner?.enabled === true) {
+      showSpinner(props.spinner.duration)
+    }
+  },
+)
```

Comment:
@jamesgeorge007 An example use case would be:-
The enabled property of the spinner is attached to a variable named `fetching` with a value of true when the request is being fetched and false once the data has been loaded. So if the data has been loaded instantaneously, fetching and enabled property concurrently becomes false. This will show the spinner in the UI for a fraction of a second and disappear which can make the UI look janky. By setting a duration, we can make sure that though fetching is turned to false immediately, the spinner will at least run for a minimum duration before disappearing away.

Reply:
Isn't the use case specific to the place of usage? Can we make it a `boolean` prop, say `loading` and the toggling can be done after a certain delay at the place of usage if necessary?